compiler/nilcheck.nim

```diff
@@ -0,0 +1,624 @@
+      var h = history(r, name)
+      assert h.len > 0
+      # echo "history", name, value, r[name], h[^1].info.line
+      result.store(name, union(value, r[name]), TAssign, h[^1].info)
+
+proc checkAsgn(target: PNode, assigned: PNode; conf, map): Check =
+  if assigned.kind != nkEmpty:
+    result = check(assigned, conf, map)
+  else:
+    result = (typeNilability(target.typ), map)
+  if result.map.isNil:
+    result.map = map
+  let t = symbol(target) # $target
+  case target.kind:
+  of nkNilLit:
+    #result.map[t] = Nil
+    result.map.store(t, Nil, TAssign, target.info)
+  else:
+    result.map.store(t, result.nilability, TAssign, target.info)
+
+proc checkReturn(n, conf, map): Check =
+  # return n same as result = n; return
+  result = check(n[0], conf, map)
+  result.map.store(resultId, result.nilability, TAssign, n.info)
+
+
+proc checkFor(n, conf, map): Check =
+  var m = map
+  var map0 = map.copyMap()
+  m = check(n.sons[2], conf, map).map.copyMap()
+  if n[0].kind == nkSym:
+    m.store(symbol(n[0]), typeNilability(n[0].typ), TAssign, n[0].info)
+  var map1 = m.copyMap()
+  var check2 = check(n.sons[2], conf, m)
+  var map2 = check2.map
+  
+  result.map = union(map0, map1)
+  result.map = union(result.map, map2)
+  result.nilability = Safe
+
+proc checkWhile(n, conf, map): Check =
+  var m = checkCondition(n[0], conf, map, false, false)
+  var map0 = map.copyMap()
+  m = check(n.sons[1], conf, m).map
+  var map1 = m.copyMap()
+  var check2 = check(n.sons[1], conf, m)
+  var map2 = check2.map
+  
+  result.map = union(map0, map1)
+  result.map = union(result.map, map2)
+  result.nilability = Safe
+
+proc checkInfix(n, conf, map): Check =
+  if n[0].kind == nkSym:
+    var mapL: NilMap
+    var mapR: NilMap
+    if n[0].sym.magic notin {mAnd, mEqRef}:
+      mapL = checkCondition(n[1], conf, map, false, false)
+      mapR = checkCondition(n[2], conf, map, false, false)
+    case n[0].sym.magic:
+    of mOr:
+      result.map = union(mapL, mapR)
+    of mAnd:
+      result.map = checkCondition(n[1], conf, map, false, false)
+      result.map = checkCondition(n[2], conf, result.map, false, false)
+    of mEqRef:
+      if n[2].kind == nkIntLit:
+        if $n[2] == "true":
+          result.map = checkCondition(n[1], conf, map, false, false)
+        elif $n[2] == "false":
+          result.map = checkCondition(n[1], conf, map, true, false)
+      elif n[1].kind == nkIntLit:
+        if $n[1] == "true":
+          result.map = checkCondition(n[2], conf, map, false, false)
+        elif $n[1] == "false":
+          result.map = checkCondition(n[2], conf, map, true, false)
+      if result.map.isNil:
+        result.map = map
+    else:
+      result.map = map
+  else:
+    result.map = map
+  result.nilability = Safe
+
+proc checkIsNil(n, conf, map; isElse: bool = false): Check =
+  result.map = newNilMap(map)
+  let value = n[1]
+  let value2 = symbol(value)
+  result.map.store(value2, if not isElse: Nil else: Safe, TArg, n.info)
```

Comment:
```suggestion
  result.map.store(symbol(n[1]), if not isElse: Nil else: Safe, TArg, n.info)
```

`value` and `value2` is only used once and they are short, dont worth the assign of new variable.

Reply:
yes